chains/evm/contracts/pools/USDC/CCTPTokenPool.sol

```diff
@@ -0,0 +1,123 @@
+import {USDCSourcePoolDataCodec} from "../../libraries/USDCSourcePoolDataCodec.sol";
+import {TokenPool} from "../TokenPool.sol";
+import {AuthorizedCallers} from "@chainlink/contracts/src/v0.8/shared/access/AuthorizedCallers.sol";
+
+import {IERC20} from "@openzeppelin/contracts@4.8.3/token/ERC20/IERC20.sol";
+
+/// @notice CCTP token pool that delegates minting and burning responsibilities of USDC to the CCTPVerifier contract.
+/// @dev This pool does not mutate the token state. It does not actually burn USDC via TokenMessenger on source or mint via MessageTransmitter on destination.
+/// It remains responsible for rate limiting and other validations while outsourcing token management to the CCTPVerfier contract.
+/// This token pool should never have a balance of USDC at any point during a transaction, otherwise funds will be lost.
+/// The caller of lockOrBurn is responsible for sending USDC to the CCTPVerifier contract instead.
+contract CCTPTokenPool is TokenPool, ITypeAndVersion, AuthorizedCallers {
+  string public constant override typeAndVersion = "CCTPTokenPool 1.7.0-dev";
+
+  constructor(
+    IERC20 token,
+    uint8 localTokenDecimals,
+    address advancedPoolHooks,
+    address rmnProxy,
+    address router,
+    address[] memory allowedCallers
+  ) TokenPool(token, localTokenDecimals, advancedPoolHooks, rmnProxy, router) AuthorizedCallers(allowedCallers) {}
+
+  /// @inheritdoc IPoolV2
+  function lockOrBurn(
+    Pool.LockOrBurnInV1 calldata lockOrBurnIn,
+    uint16 blockConfirmationRequested,
+    bytes calldata tokenArgs
+  ) public virtual override returns (Pool.LockOrBurnOutV1 memory, uint256 destTokenAmount) {
+    return (_lockOrBurn(lockOrBurnIn, blockConfirmationRequested, tokenArgs), lockOrBurnIn.amount);
+  }
+
+  /// @inheritdoc IPoolV1
+  function lockOrBurn(
+    Pool.LockOrBurnInV1 calldata lockOrBurnIn
+  ) public virtual override returns (Pool.LockOrBurnOutV1 memory lockOrBurnOutV1) {
+    return _lockOrBurn(lockOrBurnIn, WAIT_FOR_FINALITY, "");
+  }
+
+  /// @dev The _validateLockOrBurn check is an essential security check.
+  /// @dev The call to _lockOrBurn(amount) is omitted because this pool is not responsible for token management.
+  /// LockedOrBurned is still emitted for consumers that expect it.
+  function _lockOrBurn(
+    Pool.LockOrBurnInV1 calldata lockOrBurnIn,
+    uint16 blockConfirmationRequested,
+    bytes memory tokenArgs
+  ) internal virtual returns (Pool.LockOrBurnOutV1 memory lockOrBurnOutV1) {
+    _validateLockOrBurn(lockOrBurnIn, blockConfirmationRequested, tokenArgs);
+
+    emit LockedOrBurned({
+      remoteChainSelector: lockOrBurnIn.remoteChainSelector,
+      token: address(i_token),
+      sender: msg.sender,
+      amount: lockOrBurnIn.amount
+    });
+
+    return Pool.LockOrBurnOutV1({
+      destTokenAddress: getRemoteToken(lockOrBurnIn.remoteChainSelector),
+      destPoolData: USDCSourcePoolDataCodec._encodeSourceTokenDataPayloadV2WithCCV()
```

Comment:
Might be easier to parse when inlined instead of a separate function in a different file